src/service/pom.rs

```diff
@@ -152,29 +169,77 @@ impl crate::server::MyService {
                 }
             }
             // start instance
-            let process = match Command::new("bash")
-                .arg("-c")
-                .arg(entrypoint)
-                .current_dir(&protocol_package_dir)
-                .env("COLINK_CORE_ADDR", core_addr)
-                .env("COLINK_JWT", user_jwt)
-                .stdout(Stdio::null())
-                .stderr(Stdio::null())
-                .spawn()
-            {
-                Ok(child) => child,
-                Err(err) => return Err(Status::internal(err.to_string())),
+            let xid = if let Some(entrypoint) = entrypoint {
+                let mut command = Command::new("bash");
+                command
+                    .arg("-c")
+                    .arg(entrypoint)
+                    .current_dir(&protocol_package_dir)
+                    .env("COLINK_CORE_ADDR", core_addr)
+                    .env("COLINK_JWT", user_jwt)
+                    .stdout(Stdio::null())
+                    .stderr(Stdio::null());
+                if !request.get_ref().vt_public_addr.is_empty() {
+                    command.env(
+                        "COLINK_VT_PUBLIC_ADDR",
+                        request.get_ref().vt_public_addr.clone(),
+                    );
+                }
+                let process = match command.spawn() {
+                    Ok(child) => child,
+                    Err(err) => return Err(Status::internal(err.to_string())),
+                };
+                process.id().to_string()
+            } else if let Some(docker_image) = docker_image {
+                let mut args = vec![
+                    "run".to_string(),
+                    "-dit".to_string(),
+                    "--rm".to_string(),
+                    "--net=host".to_string(),
+                    "-e".to_string(),
+                    format!("COLINK_CORE_ADDR={core_addr}"),
+                    "-e".to_string(),
+                    format!("COLINK_JWT={user_jwt}"),
+                ];
+                if !request.get_ref().vt_public_addr.is_empty() {
+                    args.push("-e".to_string());
+                    args.push(format!(
+                        "COLINK_VT_PUBLIC_ADDR={}",
+                        request.get_ref().vt_public_addr
+                    ));
+                }
+                args.push(docker_image.to_string());
+                let container_id = match Command::new("docker")
+                    .args(args)
+                    .current_dir(&protocol_package_dir)
+                    .output()
+                {
+                    Ok(output) => String::from_utf8_lossy(&output.stdout).to_string(),
+                    Err(err) => return Err(Status::internal(err.to_string())),
+                };
+                container_id.replace(|c: char| !c.is_ascii_alphanumeric(), "")
+            } else {
+                unreachable!()
```

Comment:
why do we need this branch?

Reply:
`An if expression evaluates to the same value as the executed block, or () if no block is evaluated. An if expression must have the same type in all situations.`
https://doc.rust-lang.org/reference/expressions/if-expr.html#if-expressions